### ISSUE TITLE
Tabs not displayed when `cat`ing files

### ISSUE THREAD

**Issue by prometheanfire** [jwilm/alacritty#2764]
Gentoo, git master (as of 20190828-235959), Wayland.

This seems to be a thing supported in gn
I'm not sure if it's posix or not, but other terminal emulators display a 'tab' character which is useful for copying.  Alacritty does not support this (neither does kitty, but they don't care).  xfce4-terminal, gnome-terminal, urxvt supports it though.  The stackexchange link below has some useful tests for outputing tabs though.  When piping into wgetpaste it seems to work, but still doesn't work with when copying/pasting.

More info:

https://unix.stackexchange.com/questions/324676/output-tab-character-on-terminal-window/324679

https://bugzilla.gnome.org/show_bug.cgi?id=769316

**Comment by prometheanfire** [comment]
https://gist.github.com/prometheanfire/02d9f52aea5cd27222e0e49540ca8244 shows as piped to `wgetpaste`.  when displaying in the terminal tabs are just spaces.  This makes it very hard to submit patches upstream (or copying anything with tabs...).

ECMA-48 is the standard that needs to be followed.

**Comment by kchibisov** [comment]
@prometheanfire Do you mean drawing some "visual" char for tabs? If you cat file with tabs and then copy it, you'll copy everything with tabs. I've just wrote a file with tabs and then copy pasted it to vim and everything was fine (I'm also on Wayland, so we're using the same clipboard backend). The problem that you've faced probably, is that you didn't select all chars that forms a tab (default to 8).

**Comment by prometheanfire** [comment]
you are correct, copying the text then pasting it pastes as tabs (just tested the same way, though I already had a patch file with tabs).

I guess what I'm talking about is the tabs do not show up as 'tabs'. I'm not sure what method the other terminals use, but when I highlight text in the terminals that I mentioned as 'working' it'll only highlight one 'block' as the tab, instead of the 8 spaces (or whatever is configured).  I imagine they are just using a hack to make a space extend or something (the stackexchange post mentioned that it was hackish), but it is useful.

Also, does alacritty do something when text is copied out to preserve the tabs?  I noticed that kitty does not preserve tabs on copy.

All in all this is much better than I imagined :D

**Comment by kchibisov** [comment]
We do handle tabs explicitly, this was added here 0c3e2861 (related PR #991). The behavior you're referring to is related to how our selection works.

FYI tab is just left most char on this 8 spaces chunk (chunk can be less, due to how tabs work), so to select just one tab you can select left most char of this chunk.

However we can probably do something about our selection to make text with tabs selecting process more intuitive.

**Comment by chrisduerr** [comment]
I thought we already had an issue for this, but I guess we do not. We already do something very similar with double-width characters, so I'd assume similar logic would apply here. I'm just not sure how easy it will be to get the length of a tab (since it's not always the same number of cells), but I think we should have access to that where we expand the selection for double-width cells.

**Comment by pwnorbitals** [comment]
It looks like we had some progress on this issue. Is it solved on latest builds ?

**Comment by kchibisov** [comment]
it's not, you may pick up the dropped PR if you want.

**Comment by Dr-McFish** [comment]
Hello, I want to try my hand at this. What went wrong or what remains to be done in the dropped PR? I have read thorough the PR and surrounding conversation, and it wasn't clear to me. I also tried downloading and compiling the old PR, and didn't quite succeed. 

Because it is so old, not sure which rust compiler version to use. The current stable compiler breaks a dependency. I also tried 1.44.0 compiles but directly crashes: ```interface 'wl_output' has no event 4
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 0, kind: Other, message: "Success" }', /home/drmcfish/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.22.1/src/platform_impl/linux/wayland/event_loop.rs:337:19```
